src/MIDebugEngine/Engine.Impl/DebugUnixChildProcess.cs

```diff
@@ -0,0 +1,260 @@
+        private int _newpid;
+        private LaunchOptions _launchOptions;
+        private string _mainBreak;
+
+        public DebugUnixChild(DebuggedProcess process, LaunchOptions launchOptions)
+        {
+            _process = process;
+            _state = State.Init;
+            _launchOptions = launchOptions;
+        }
+
+        private async Task ProcessChild()
+        {
+            Debug.Assert(_newpid != 0, "Child process id not found.");
+            if (_newpid != 0)
+            {
+                int inf = _process.InferiorByPid(_newpid);
+                if (inf != 0)
+                {
+                    await _process.ConsoleCmdAsync("inferior " + inf.ToString());
+                    if (!string.IsNullOrEmpty(_mainBreak))
+                    {
+                        await _process.MICommandFactory.BreakDelete(_mainBreak);
+                        _mainBreak = null;
+                    }
+                    _state = State.AtSignal;
+                    await _process.MICommandFactory.Signal("SIGSTOP");  // stop the child
+                }
+            }
+        }
+
+        private async Task RunChildToMain()
+        {
+            Debug.Assert(_newpid != 0, "Child process id not found.");
+            if (_newpid != 0)
+            {
+                int inf = _process.InferiorByPid(_newpid);
+                if (inf != 0)
+                {
+                    await _process.ConsoleCmdAsync("inferior " + inf.ToString());
+                    await SetBreakAtMain();
+                    _state = State.AtExec;
+                    await _process.MICommandFactory.ExecContinue();  // run the child
+                }
+            }
+        }
+
+        private async Task ContinueTheChild()
+        {
+            Debug.Assert(_state == State.AtExec, "wrong vfork processing state");
+            if (_newpid != 0)
+            {
+                int inf = _process.InferiorByPid(_newpid);
+                if (inf != 0)
+                {
+                    await _process.ConsoleCmdAsync("inferior " + inf.ToString());
+                    await _process.MICommandFactory.ExecContinue();  // run the child
+                }
+            }
+        }
+
+        private async Task<bool> DetachAndContinue()
+        {
+            int inf = _process.InferiorByPid(_newpid);
+            if (inf == 0)
+                return false;    // cannot process the child
+            await _process.ConsoleCmdAsync("inferior " + inf.ToString());
+            await _process.MICommandFactory.TargetDetach();     // detach from the child
+            await _process.ConsoleCmdAsync("inferior 1");
+            lock(AD7Engine.ChildProcessLaunch)
+            {
+                AD7Engine.ChildProcessLaunch.Add(_newpid);
+            }
+            string engineName;
+            Guid engineGuid;
+            _process.Engine.GetEngineInfo(out engineName, out engineGuid);
+            _launchOptions.BaseOptions.ProcessId = _newpid;
```

Comment:
This seems like a race here. Seems like we need to clone the original launch options and change these.